db/revtree.go

```diff
@@ -149,29 +173,51 @@ func (tree *RevTree) UnmarshalJSON(inputjson []byte) (err error) {
 				info.BodyKey = bodyKey
 			}
 		}
-		if rep.Channels != nil {
-			info.Channels = rep.Channels[i]
-		}
 		parentIndex := rep.Parents[i]
 		if parentIndex >= 0 {
 			info.Parent = rep.Revs[parentIndex]
 		}
 		(*tree)[revid] = &info
 	}
-	if rep.Deleted != nil {
-		for _, i := range rep.Deleted {
+
+	for _, i := range rep.Deleted {
+		info := (*tree)[rep.Revs[i]]
+		info.Deleted = true // because tree[rep.Revs[i]].Deleted=true is a compile error
+		(*tree)[rep.Revs[i]] = info
+	}
+
+	for _, i := range rep.HasAttachments {
+		info := (*tree)[rep.Revs[i]]
+		info.HasAttachments = true
+		(*tree)[rep.Revs[i]] = info
+	}
+
+	if len(rep.ChannelsMap) > 0 || len(rep.Channels_Old) > 0 {
+		leaves := tree.Leaves()
+		for iStr, channels := range rep.ChannelsMap {
+			i, err := strconv.ParseInt(iStr, 10, 64)
+			if err != nil {
+				return err
+			}
 			info := (*tree)[rep.Revs[i]]
-			info.Deleted = true // because tree[rep.Revs[i]].Deleted=true is a compile error
-			(*tree)[rep.Revs[i]] = info
+			if _, isLeaf := leaves[info.ID]; isLeaf {
```

Comment:
Do we need the leaf check here, if we're only ever marshalling ChannelsMap for leaves?